### PR TITLE
Add database based request logging

### PR DESCRIPTION
**Pull request by oscarwyatt**
Adds database based request logging, including information in key decision points in the system. This will enable us to analyse the performance of the app and the decisions it makes so that we will be able to inspect the decisions it makes and iterate it in the future

Trello: https://trello.com/c/ftSyPF2I/222-add-database-to-tagging-suggester-and-log-output

### CODE REVIEW

File: alembic/versions/9f8ce88d37e9_create_requests_table.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def upgrade():
+    op.create_table(
+        'requests',
+        sa.Column('id', sa.Integer, primary_key=True),
+        sa.Column('created_at', sa.DateTime, server_default=sa.func.current_timestamp()),
+        sa.Column('prediction_duration', sa.String),
+        sa.Column('edition_id', sa.Integer),
+        sa.Column('content', sa.Text),
```
Comment (karlbaker02) [inline alembic/versions/9f8ce88d37e9_create_requests_table.py]: Might this contain draft text which is sensitive? i.e. budgets to be announced? If so, should we be storing this?

File: config.py [hunk 2]
```diff
@@ -0,0 +1 @@
+DATABASE_URI = "postgres+psycopg2://taggingsuggester@localhost:5432/tagging_suggester_development"
```
Comment (karlbaker02) [inline config.py]: We might need to add another database here for production?

File: alembic/versions/9f8ce88d37e9_create_requests_table.py (hunk 1 above)
Comment (karlbaker02) [reply alembic/versions/9f8ce88d37e9_create_requests_table.py]: I have also been looking around to determine how SQLAlchemy maps it's `Text` type to Postgres and whether there it specifies a max length on the column it creates which might impact us, depending how much content is stored. From what I can tell it doesn't add a max length, which seems to be supported by https://github.com/sqlalchemy/sqlalchemy/issues/4443. 1 GB looks to be the maximum column size supported by Postgres for text types - https://www.postgresql.org/docs/current/datatype-character.html.

File: src/__init__.py [hunk 3]
```diff
@@ -1,18 +1,43 @@
...
+from sqlalchemy import create_engine
+from sqlalchemy.orm import sessionmaker
 
 app = Flask(__name__)
 
 tagging_suggester = TaggingSuggester()
 
+@app.before_request
+def before_request():
+    engine = create_engine(app_config.database_url(), strategy='threadlocal')
```
Comment (karlbaker02) [inline src/__init__.py]: Will this create a connection on every request? Ideally we'd store a connection that we can re-use...

File: src/__init__.py [hunk 4]
```diff
@@ -1,18 +1,43 @@
...
+    Session = sessionmaker(bind=engine)
+    g.db_session = Session()
+
 @app.route('/create', methods=['POST'])
 def tagging_suggestion():
     json = request.get_json()
-    print(json)
-    result = tagging_suggester.predict(json['text'])
+    request_record = Request(
+        created_at = datetime.now().isoformat(),
```
Comment (karlbaker02) [inline src/__init__.py]: Is this value consistent with what the database expects (a string in the format `YYYY-MM-DD`)?

Comment (karlbaker02) [reply src/__init__.py]: Also not a requirement to be specified as you have opted for a default in the creation of the table, if you didn't want to specify it.

File: src/__init__.py [hunk 5]
```diff
@@ -1,18 +1,43 @@
...
+        edition_id = json['edition_id'],
+        content = json['text'],
+        branch_predictor_probabilities = "",
```
Comment (karlbaker02) [inline src/__init__.py]: Is this meant to be empty?

File: src/models/apex_node_predictor.py [hunk 6]
```diff
@@ -73,10 +73,14 @@ def predict(self, tree, text_to_predict, translated_tokenized_text):
...
-        return sorted(results, key=results.get, reverse=True)
+        for i, probability in enumerate(class_probabilities):
+            predicted_node_content_id = model.classes_[i]
+            is_above_threshold = probability > self.threshold()
+            all_results.append({ 'node_content_id': predicted_node_content_id, 'probability': probability, 'above_threshold': is_above_threshold })
+            if is_above_threshold:
+                results[tree.find(predicted_node_content_id)] = predicted_node_content_id
+        request_record.apex_node_predictor_probabilities = str(all_results)
+        # Order results by probability (highest first)
+        return [sorted(results, key=results.get, reverse=True), request_record]
```
Comment (karlbaker02) [inline src/models/apex_node_predictor.py]: Just to confirm, we now want to return all results, including those that haven't met the prediction threshold?

File: src/models/branch_predictor.py [hunk 7]
```diff
@@ -66,21 +66,26 @@ def can_make_prediction_for_node(self, node):
...
         while self.can_make_prediction_for_node(node):
             model = dirs.open_pickle_file(self.model_path(node))
             vectorizer = dirs.open_pickle_file(self.vectorizer_path(node))
             transformed_text = vectorizer.transform([text_to_predict])
             words_to_explain_choice = nlp.prediction_explanation(vectorizer, transformed_text, translated_tokenized_text_to_predict)
             probabilities = model.predict_proba(transformed_text)
             highest_probability_content_id = model.classes_[np.argmax(probabilities)]
-            result.append({'taxon_content_id': highest_probability_content_id, 'explanation': words_to_explain_choice})
+            probabilities_for_taxon = {}
+            for i, probability in enumerate(probabilities[0]):
```
Comment (karlbaker02) [inline src/models/branch_predictor.py]: What is `probabilities[0]` here? From what I've read `predict_proba` isn't ordering these by most>least likely...

File: src/utils/app_config.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import os
+
+def database_url():
+    database_url = os.getenv("TAGGING-SUGGESTER-DATABASE-URL")
```
Comment (karlbaker02) [inline src/utils/app_config.py]: Is this meant to be different to `DATABASE_URI` specified in `config.py`?

File: src/utils/app_config.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+    if database_url is None:
+        database_url = "postgres://taggingsuggester@localhost:5432/tagging_suggester_development"
```
Comment (karlbaker02) [inline src/utils/app_config.py]: Again, do we need another database specified here for production?

Comment (oscarwyatt) [reply src/utils/app_config.py]: I don't think so - AFAIK sqlalchemy doesn't have a concept of environments

File: src/utils/app_config.py (hunk 8 above)
Comment (oscarwyatt) [reply src/utils/app_config.py]: Removed config.py

File: src/__init__.py (hunk 3 above)
Comment (karlbaker02) [reply src/__init__.py]: As discussed

File: src/__init__.py (hunk 4 above)
Comment (karlbaker02) [reply src/__init__.py]: As discussed

File: src/__init__.py (hunk 5 above)
Comment (karlbaker02) [reply src/__init__.py]: As discussed

File: src/models/branch_predictor.py (hunk 7 above)
Comment (karlbaker02) [reply src/models/branch_predictor.py]: As discussed

File: src/utils/app_config.py (hunk 9 above)
Comment (karlbaker02) [reply src/utils/app_config.py]: As discussed

File: src/utils/app_config.py (hunk 8 above)
Comment (karlbaker02) [reply src/utils/app_config.py]: 👍

File: src/models/apex_node_predictor.py (hunk 6 above)
Comment (karlbaker02) [reply src/models/apex_node_predictor.py]: As discussed

File: alembic/versions/9f8ce88d37e9_create_requests_table.py (hunk 1 above)
Comment (karlbaker02) [reply alembic/versions/9f8ce88d37e9_create_requests_table.py]: As discussed

**Review by sophiebrown01 — APPROVED** [review]
Looks good 👍 Changes discussed in person

**Pull request merged**